File: Data Structures & Algorithms/course-schedule/submission-1.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = defaultdict(list)
        for a, b in prerequisites:
            graph[a].append(b)
        visited = [0] * (numCourses)
        def dfs(i):
            if visited[i] == 2:
                return True
            if visited[i] == 1:
                return False
            visited[i] = 1
            for u in graph[i]:
                if dfs(u) == False:
                    return False
            visited[i] = 2
            return True
        for i in range(numCourses):
            if dfs(i) == False:
                return False
        return True

File: Data Structures & Algorithms/course-schedule-ii/submission-1.java
class Solution {
    private boolean dfs(ArrayList<Integer> order, int node, List<List<Integer>> graph, int[] state) {
        if (state[node] == 1) return false;    // cycle found
        if (state[node] == 2) return true;     // already finished

        state[node] = 1; // mark visiting

        for (int neigh : graph.get(node)) {
            if (!dfs(order, neigh, graph, state))
            {
                return false;
            }
        }

        state[node] = 2; // mark done
        order.add(node);
        return true;
    }

    public int[] findOrder(int numCourses, int[][] prerequisites) {
        ArrayList<Integer> order = new ArrayList<>();
        // build adjacency list
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) {
            graph.add(new ArrayList<>());
        }

        for (int[] pre : prerequisites) {
            graph.get(pre[0]).add(pre[1]);
        }

        int[] state = new int[numCourses]; // 0 = unvisited, 1 = visiting, 2 = visited

        for (int i = 0; i < numCourses; i++) {
            if (state[i] == 0) {
                if (!dfs(order, i, graph, state))
                {
                    return new int[0];
                }
            }
        }
        int[] res = new int[numCourses];
        for(int i = 0; i < order.size(); i++)
        {
            res[i] = order.get(i);
        }
        return res;
    }
}

